Adopt dynamicDowncast<> in ImageLoader and RenderImage

annevk opened the pull request:
#### d4c1f6993b5637aef5023927054a1f8e0d95eb4f
<pre>
Adopt dynamicDowncast&lt;&gt; in ImageLoader and RenderImage
<a href="https://bugs.webkit.org/show_bug.cgi?id=267209">https://bugs.webkit.org/show_bug.cgi?id=267209</a>

Reviewed by Chris Dumez.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::renderImageResource):
(WebCore::ImageLoader::decode):
(WebCore::ImageLoader::imageVisibleInViewport const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::RenderImage):
(WebCore::RenderImage::notifyFinished):
(WebCore::isDeferredImage):
(WebCore::RenderImage::paintAreaElementFocusRing):
(WebCore::RenderImage::paintIntoRect):
(WebCore::RenderImage::updateAltText):
(WebCore::RenderImage::computeIntrinsicRatioInformation const):
(WebCore::RenderImage::embeddedContentBox const):

Canonical link: <a href="https://commits.webkit.org/272781@main">https://commits.webkit.org/272781@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/573f2959ec2163b12d448c8d4cfc9af0cde9a8e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33019 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11792 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34923 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35659 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29837 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33990 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14133 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8962 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29268 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33494 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9938 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29502 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8664 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/8819 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29460 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36995 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29995 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29869 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/34946 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8954 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6895 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32801 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10646 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7655 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9545 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9587 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -206,12 +206,11 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
-            options.referrerPolicy = imageElement.referrerPolicy();
-            options.fetchPriorityHint = imageElement.fetchPriorityHint();
-            if (imageElement.usesSrcsetOrPicture())
+        RefPtr imageElement = dynamicDowncast<HTMLImageElement>(element());
+        if (imageElement) {
```
```suggestion
        if (RefPtr imageElement = dynamicDowncast<HTMLImageElement>(element())) {
```

annevk replied:
That puts it out-of-scope when it's needed later on.

annevk commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -811,10 +807,10 @@ void RenderImage::updateAltText()
+    if (auto input = dynamicDowncast<HTMLInputElement>(*element()))
+        m_altText = input->altText();
+    else if (auto image = dynamicDowncast<HTMLImageElement>(*element()))
+        m_altText = image->altText();
```
RefPtr?

cdumez replied:
Just for getting the altText(). It may not be super critical. But I do think we should either use `auto*` here to `RefPtr` rather than `auto`.

annevk commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -905,10 +900,10 @@ bool RenderImage::needsPreferredWidthsRecalculation() const
-        return downcast<SVGImage>(*cachedImage->image()).embeddedContentBox();
-
+    if (auto* cachedImage = this->cachedImage()) {
+        if (auto image = dynamicDowncast<SVGImage>(cachedImage->image()))
```
auto*

cdumez commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -528,14 +526,12 @@ void ImageLoader::decode()
 
-    Image* image = m_image->image();
-    if (!is<BitmapImage>(image)) {
+    auto* bitmapImage = dynamicDowncast<BitmapImage>(m_image->image());
```
Would be nice to use a RefPtr here given we call `decode()` on this later.

cdumez commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -397,8 +397,8 @@ void RenderImage::notifyFinished(CachedResource& newImage, const NetworkLoadMetr
 
-    if (is<HTMLImageElement>(element()))
-        page().didFinishLoadingImageForElement(downcast<HTMLImageElement>(*element()));
+    if (auto* image = dynamicDowncast<HTMLImageElement>(element()))
```
Seems like a RefPtr here would be useful.

cdumez commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -702,10 +699,9 @@ ImageDrawResult RenderImage::paintIntoRect(PaintInfo& paintInfo, const FloatRect
-    if (is<BitmapImage>(image))
-        downcast<BitmapImage>(*image).updateFromSettings(settings());
+    auto* image = imageResource().image().get();
+    if (auto* bitmapImage = dynamicDowncast<BitmapImage>(image))
```
A RefPtr would be nice here.

cdumez commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -811,10 +807,10 @@ void RenderImage::updateAltText()
-        m_altText = downcast<HTMLInputElement>(*element()).altText();
-    else if (is<HTMLImageElement>(*element()))
-        m_altText = downcast<HTMLImageElement>(*element()).altText();
+    if (auto input = dynamicDowncast<HTMLInputElement>(*element()))
```
auto* ?

cdumez reviewed (approved):
LGTM with nits.

annevk committed:
Adopt dynamicDowncast<> in ImageLoader and RenderImage
https://bugs.webkit.org/show_bug.cgi?id=267209

Reviewed by Chris Dumez.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::renderImageResource):
(WebCore::ImageLoader::decode):
(WebCore::ImageLoader::imageVisibleInViewport const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::RenderImage):
(WebCore::RenderImage::notifyFinished):
(WebCore::isDeferredImage):
(WebCore::RenderImage::paintAreaElementFocusRing):
(WebCore::RenderImage::paintIntoRect):
(WebCore::RenderImage::updateAltText):
(WebCore::RenderImage::computeIntrinsicRatioInformation const):
(WebCore::RenderImage::embeddedContentBox const):

Canonical link: https://commits.webkit.org/272781@main
